src/Framework/Hosting.AspNetCore/Hosting/DotvvmHealthCheck.cs

```diff
@@ -0,0 +1,58 @@
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using DotVVM.Framework.Compilation;
+using DotVVM.Framework.Utils;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+
+namespace DotVVM.Framework.Hosting
+{
+    public class DotvvmHealthCheck : IHealthCheck
+    {
+        private readonly IDotvvmViewCompilationService compilationService;
+
+        public DotvvmHealthCheck(IDotvvmViewCompilationService compilationService)
+        {
+            this.compilationService = compilationService;
+        }
+
+        public Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
+        {
+            var c = compilationService.GetRoutes().AddRange(compilationService.GetControls().AddRange(compilationService.GetMasterPages()));
+
+            if (!c.Any(p => p.Status == CompilationState.CompilationFailed))
+            {
+                return Task.FromResult(new HealthCheckResult(HealthStatus.Healthy));
+            }
+
+            // if nothing compiles, we say unhealthy. If at least something works, we say only degraded
+            var state =
+                c.Any(p => p.Status is CompilationState.CompletedSuccessfully or CompilationState.CompilationWarning)
+                    ? HealthStatus.Degraded
+                    : HealthStatus.Unhealthy;
+            
+            var views = c.Where(p => p.Status == CompilationState.CompilationFailed)
+                         .Select(c => c.VirtualPath);
+            return Task.FromResult(new HealthCheckResult(state, $"Dothtml pages can not be compiled: {views.Take(10).StringJoin(", ")}. See /_dotvvm/diagnostics/compilation for more information."));
```

Comment:
The diagnostic page url should not be hardcoded since it can be configured in `DotvvmCompilationPageConfiguration.Url`.